target_snowflake/sinks.py

```diff
@@ -1,44 +1,140 @@
+        )
+
+    def _load_csv_files_from_table_stage(self, full_table_name: str) -> None:
+        """Loads all files from the table's built-in stage."""
+        self.connector.connection.execute(
+            f"COPY INTO {full_table_name} "
+            f"FILE_FORMAT = (TYPE = CSV FIELD_DELIMITER = '\\t' SKIP_HEADER = 1);"
+        )
+
+    @property
+    def timestamp_time(self) -> datetime.datetime:
+        try:
+            return self._timestamp_time
+        except AttributeError:
+            self._timestamp_time = datetime.datetime.now(
+                tz=pytz.timezone(self.config["timestamp_timezone"])
+            )
+            return self._timestamp_time
+
+    @property
+    def filepath_replacement_map(self) -> Dict[str, str]:
+        return {
+            "stream_name": self.stream_name,
+            "datestamp": self.timestamp_time.strftime(self.config["datestamp_format"]),
+            "timestamp": self.timestamp_time.strftime(self.config["timestamp_format"]),
+        }
+
+    def _tmp_dir(self) -> Path:
+        """Return a temp directory path."""
+        return Path(tempfile.mkdtemp(prefix="target-snowflake"))
+
+    @property
+    def destination_path(self) -> Path:
+        result = str(self._tmp_dir / self.config["file_naming_scheme"])
+        for key, val in self.filepath_replacement_map.items():
+            replacement_pattern = "{" f"{key}" "}"
+            if replacement_pattern in result:
+                result = result.replace(replacement_pattern, val)
+
+        if self.config.get("output_path_prefix", None) is not None:
+            result = f"{self.config['output_path_prefix']}{result}"
+
+        return Path(result)
 
     def process_batch(self, context: dict) -> None:
-        super().process_batch(context)
-        self.logger.info(f"TODO: load '{self.destination_path}' into Snowflake...")
+        """Write out any prepped records and return once fully written."""
+        records: List[Dict[str, Any]] = context["records"]
+        if "record_sort_property_name" in self.config:
+            sort_property_name = self.config["record_sort_property_name"]
+            records = sorted(records, key=lambda x: x[sort_property_name])
+
+    def bulk_insert_records(
+        self, full_table_name: str, schema: dict, records: Iterable[Dict[str, Any]]
+    ) -> Optional[int]:
+        if not records:
+            self.logger.warning(f"No values in {self.stream_name} records collection.")
+            return 0
+
+        output_file: Path = self.destination_path
+        self.logger.info(f"Writing to destination file '{output_file.resolve()}'...")
 
-        # copy the file to the stage
-        # load the file from the stage into the database
-        # (optional) archive the file
+        self._write_csv(output_file, records)
+        self._upload_csv_to_table_stage(output_file, full_table_name)
+        self._load_csv_files_from_table_stage(full_table_name)
```

Comment:
One thing I wanted to try and work into this target was smarter handling of connections. Snowflake warehouses can be spun down after a period of inactivity, but only when all the connections are closed. If you spend most of your time writing to a local CSV file, you're paying to keep that server running for no reason.

Another thing I ran into was connection timeouts. The Snowflake connection times out after a period of 4 hours. I had a tap for an API with very aggressive throttling regularly time out writing until we lowered the batch size to something tiny. Ideally We'd connect early and set up all the table schemas, start loading the data into CSVs letting the connection close after a few minutes of inactivity, and reconnecting when it's time to run `PUT`/`COPY INTO`(/`ACTIVATE_VERSION`)